File: android/src/main/kotlin/com/example/device_info_plugin/DeviceInfoPlugin.kt
package com.example.device_info_plugin

import android.os.Build
import io.flutter.embedding.engine.plugins.FlutterPlugin

class DeviceInfoPlugin : FlutterPlugin, DeviceInfoHostApi {

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        DeviceInfoHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        DeviceInfoHostApi.setUp(binding.binaryMessenger, null)
    }

    override fun getDeviceInfo(): DeviceInfoMessage {
        return DeviceInfoMessage(
            osVersion = "Android ${Build.VERSION.RELEASE}",
            deviceModel = Build.MODEL
        )
    }
}

File: ios/Classes/DeviceInfoPlugin.swift
import Flutter
import UIKit

public class DeviceInfoPlugin: NSObject, FlutterPlugin, DeviceInfoHostApi {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let messenger = registrar.messenger()
    let api = DeviceInfoPlugin()
    DeviceInfoHostApiSetup.setUp(binaryMessenger: messenger, api: api)
  }

  func getDeviceInfo() throws -> DeviceInfoMessage {
    var message = DeviceInfoMessage()
    message.osVersion = "iOS " + UIDevice.current.systemVersion
    message.deviceModel = UIDevice.current.model
    return message
  }
}
